### DISCUSSION TITLE
RJMS vs AMQP Destinations without JNDI

### DISCUSSION THREAD

**Discussion by ahensley-trc** [rabbitmq/rabbitmq-jms-client#531] [Q&A] [answered]
Hi folks,

We have been looking at utilizing the JMS client with an application that doesn't look up a JMS `Destination`, only the `ConnectionFactory` is looked up in JNDI. The app calls `Session.createQueue()` directly, unfortunately. Since we need the client to convert AMQP messages put on the RabbitMQ queue to JMS messages, I think we are probably out of luck here, right? I have reviewed the documentation and looked at the code, and I don't see a way to set `amqp` to `true` when the `Queue` is created this way.

Essentially I am looking for confirmation that I haven't missed a way to do this. For example, the Apache Qpid JMS client [defines a binding URL scheme](https://qpid.apache.org/releases/qpid-jms-amqp-0-x-6.4.0/jms-amqp-0-8-book/JMS-Client-0-8-Client-Understanding-Session.html#JMS-Client-0-8-Client-Understanding-Session-CreateQueue) that can be passed as the queue name to `createQueue()`, allowing AMQP properties like exchange name and routing key to be set that way. (The Qpid client won't work for us for unrelated reasons). 

So in short, is it the case that if a JMS client does not get the `Destination` from the JNDI context, there is no way to set it to an AMQP `Destination`?

Thanks very much for the help,
Alan

**Comment by acogoluegnes** [comment, marked as answer]
You can use the `RMQDestination` constructor and the client library will declare the appropriate topology. See the documentation:
https://rabbitmq.github.io/rabbitmq-jms-client/3.x/snapshot/htmlsingle/index.html#jms-amqp-rmqdestination-constructor
https://rabbitmq.github.io/rabbitmq-jms-client/3.x/snapshot/htmlsingle/index.html#jms-amqp-0-9-1-destination-definitions

**Reply by ahensley-trc** [reply]
Thanks for the quick reply Arnaud. I think in our case, since we don't have control over the application code and can't re-build it to use any of the RabbitMQ JMS client classes directly, we couldn't use this approach. We're only able to change the JMS client configuration settings exposed by the app (things like initial context factory, queue name, ConnectionFactory class name, etc.).

**Reply by acogoluegnes** [reply]
OK. There is nothing much we can do in this case, `Session#create` creates a regular queue, it won't try to make it interoperable between a JMS client and AMQP clients.

**Reply by ahensley-trc** [reply]
Ok, that is what it looked like to me. I appreciate the confirmation. Thanks again!
